opendps/func_settings.c

```diff
@@ -0,0 +1,681 @@
+ */
+static set_param_status_t set_parameter(char *name, char *value)
+{
+    return ps_unknown_name;
+}
+
+/**
+ * @brief      Get function parameter
+ *
+ * @param[in]  name       name of parameter
+ * @param[in]  value      value of parameter as a string - always in SI units
+ * @param[in]  value_len  length of value buffer
+ *
+ * @retval     set_param_status_t status code
+ */
+static set_param_status_t get_parameter(char *name, char *value, uint32_t value_len) {
+    return ps_unknown_name;
+}
+
+
+static int32_t get_v_adc_k() {
+    return v_adc_k_coef * 1000;
+}
+static void set_v_adc_k(ui_number_t *item) {
+    v_adc_k_coef = item->value / 1000.0f;
+    if (item->color == RED) v_adc_k_coef = - v_adc_k_coef;
+}
+static int32_t get_v_adc_c() {
+    return v_adc_c_coef * 1000;
+}
+static void set_v_adc_c(ui_number_t *item) {
+    v_adc_c_coef = item->value / 1000.0f;
+    if (item->color == RED) v_adc_c_coef = - v_adc_c_coef;
+}
+
+
+static int32_t get_v_dac_k() {
+    return v_dac_k_coef * 1000;
+}
+static void set_v_dac_k(ui_number_t *item) {
+    v_dac_k_coef = item->value / 1000.0f;
+    if (item->color == RED) v_dac_k_coef = - v_dac_k_coef;
+}
+static int32_t get_v_dac_c() {
+    return v_dac_c_coef * 1000;
+}
+static void set_v_dac_c(ui_number_t *item) {
+    v_dac_c_coef = item->value / 1000.0f;
+    if (item->color == RED) v_dac_c_coef = - v_dac_c_coef;
+}
+
+
+
+static int32_t get_a_adc_k() {
+    return a_adc_k_coef * 1000;
+}
+static void set_a_adc_k(ui_number_t *item) {
+    a_adc_k_coef = item->value / 1000.0f;
+    if (item->color == RED) a_adc_k_coef = - a_adc_k_coef;
+}
+static int32_t get_a_adc_c() {
+    return a_adc_c_coef * 1000;
+}
+static void set_a_adc_c(ui_number_t *item) {
+    a_adc_c_coef = item->value / 1000.0f;
+    if (item->color == RED) a_adc_c_coef = - a_adc_c_coef;
+}
+
+
+static int32_t get_a_dac_k() {
+    return a_dac_k_coef * 1000;
+}
+static void set_a_dac_k(ui_number_t *item) {
+    a_dac_k_coef = item->value / 1000.0f;
+    if (item->color == RED) a_dac_k_coef = - a_dac_k_coef;
+}
+static int32_t get_a_dac_c() {
+    return a_dac_c_coef * 1000;
+}
+static void set_a_dac_c(ui_number_t *item) {
+    a_dac_c_coef = item->value / 1000.0f;
+    if (item->color == RED) a_dac_c_coef = - a_dac_c_coef;
+}
+
+
+static int32_t get_vin_adc_k() {
+    return vin_adc_k_coef * 1000;
+}
+static void set_vin_adc_k(ui_number_t *item) {
+    vin_adc_k_coef = item->value / 1000.0f;
+    if (item->color == RED) vin_adc_k_coef = - vin_adc_k_coef;
+}
+static int32_t get_vin_adc_c() {
+    return vin_adc_c_coef * 1000;
+}
+static void set_vin_adc_c(ui_number_t *item) {
+    vin_adc_c_coef = item->value / 1000.0f;
+    if (item->color == RED) vin_adc_c_coef = - vin_adc_c_coef;
+}
+
+
+static int32_t get_brightness() {
+    return (hw_get_backlight() / 1.28f) * 1000;
+}
+static void set_brightness(ui_number_t *item) {
+    if (item->value > 100 * 1000) item->value = 100 * 1000;
+    if (item->value < 0) item->value = 0;
```

Comment:
Good suggestion. 

Unfortunately, the way I have the screen set up is to have 6 static fields that are used for each 'page'. Because each field might have a different min/max, I can't actually set a hard min/max value. Thus, the check is currently done as part of the set function callback.

Perhaps when I have time, or if someone else is willing to implement, the min/max can be updated based on a lookup table or callback function when set_page() is called and each field is updated.

Reply:
Using struct containing min/max values for each field.